mining: add cooldown to createNewBlock() immediately after IBD

Sjors opened the pull request:
As reported in #33994, connected mining clients will receive a flood of new templates if the node is still going through IBD or catching up on the last 24 hours. This PR fixes that using an _optional_ cooldown mechanism, only applied to `createNewBlock()`.

First, cooldown waits for IBD. Then, as the tip keeps moving forward, it waits a few seconds to see if the tip updated. If so, it restarts the timer and waits again. The trade-offs for this mechanism are explained below.

Because this PR changes `createNewBlock()` from a method that returns quickly to one that can block for minutes, we rely on #34568 to fix a bug in our `.capnp` definition, adding the missing `context` to `createNewBlock` (and `checkBlock`).

The second commit then adds an `interrupt()` method so that clients can cleanly disconnect.

---

## Rationale

The cooldown argument is optional, and not used by internal non-IPC code, for two reasons:

1. The mechanism wreaks havoc on the functional test suite, which would require very careful mock time handling to work around. But that's pointless, because only IPC clients need it.
2. It needs to be optional for IPC clients too, because in some situations, like a signet with only one miner, waiting for IBD can mean being stuck forever.

The reason it's only applied to `createNewBlock()` is that this is the first method called by clients; `waitNext()` is a method on the interface returned by `createNewBlock()`, at which point the cooldown is done.

After IBD, we wait N seconds if the header is N blocks ahead of the tip, with a minimum of 3 and a maximum of 20 seconds. The minimum waiting time is short enough that it shouldn't be annoying or confusing for someone manually starting up a client. While the maximum should be harmless if it happens spuriously (which it shouldn't).

If the minimum wait is too short, clients get a burst of templates, as observed in the original issue. We can't entirely rule this out without a lot of additional complexity (like scanning our own log file for heuristics). This PR should make it a lot less likely, and thanks to the IBD wait also limit it to one day worth of blocks (`-maxtipage`).

Some test runs on an M4 MacBook Pro, where I had a node catch up on the last few days worth of blocks:

<img width="872" height="972" alt="Scherm­afbeelding 2026-02-04 om 18 21 17" src="https://github.com/user-attachments/assets/7902a0f2-0e0b-4604-9688-cec2da073261" />

As the chart shows, sometimes it takes longer than 3 seconds. But it turns out that in all those cases there were quite a few headers ahead of the tip. It also demonstrates that it's important to first wait for IBD, because it's less likely a random tip update takes longer than 20 seconds.

- modified sv2-apps: https://github.com/Sjors/sv2-apps/tree/2026/02/cooldown
- test script: https://gist.github.com/Sjors/feb6122c97acc2b9e6d66b168614609c#file-run_mainnet_pool_loop-zsh
- chart script: https://gist.github.com/Sjors/feb6122c97acc2b9e6d66b168614609c#file-tip_interval_charts-py

bensig commented:
ACK e8b01c43bb05f1a52adf1279280a744111b8dc45

Tested on macOS - `interface_ipc.py` passes.

Code review:
- Using `SteadyClock` for the cooldown (independent of mocktime) is the right call
- `BestHeaderAheadOfActiveTip()` correctly uses ancestor check to ignore competing branches
- 1s timeout prevents DoS from header-only attacks while still allowing catch-up during final IBD
- 

Nice fix for #33994

w0xlt reviewed (commented):
Approach ACK

w0xlt reviewed (commented):
ACK https://github.com/bitcoin/bitcoin/pull/34184/commits/9817657d463ff1d8502a5edf0641e88b9b1d4853

ryanofsky reviewed (approved):
Code review ACK 9817657d463ff1d8502a5edf0641e88b9b1d4853. This seems like probably the simplest change that can resolve #33994.

I guess I have two lingering concerns about it:

- This can cause the createNewBlock function to block for potentially a very long time when it didn't before and createNewBlock doesn't currently have any cancellation option other than shutting down the node. So it could make it hard for IPC clients to disconnect cleanly, and I suspect we may want to follow up here by adding some timeout or cancellation mechanism.

- I'm not very sure about effectiveness of the cooldown strategy, and wouldn't be surprised if there were cases where it didn't work well but it does seem like a good starting point, and pretty conservative.

Sjors commented on test/functional/interface_ipc.py:
```diff
@@ -265,19 +271,38 @@ async def async_routine():
+                # Lower-bound only: a heavily loaded CI host might still exceed 0.9s
+                # even without the cooldown, so this can miss regressions but avoids
+                # spurious failures.
+                assert_greater_than_or_equal(time.time() - start, 0.9)
```
See #34452. No preference on merge order, either should be easy to rebase.

ryanofsky reviewed (commented):
Code review b57da688c21233e852c5afe5c5afafd3cae84e91. Thanks for the updates! Documentation update and while loop simplification since last review look good. I'm concerned about the new `MinerImpl::m_interrupt` variable, however, since I can see it leading to segfaults if clients destroy the `Mining` interface too early (even if they don't call the `interrupt` method). So I suggested moving the variable to the `KernelNotications` class to avoid this.

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -995,6 +1000,7 @@ class MinerImpl : public Mining
     NodeContext* context() override { return &m_node; }
     ChainstateManager& chainman() { return *Assert(m_node.chainman); }
     KernelNotifications& notifications() { return *Assert(m_node.notifications); }
+    bool m_interrupt{false};
```
I mostly just cloned the pattern we used for `waitNext()`. Does it have a similar issue?

ryanofsky reviewed (commented):
Code review b57da688c21233e852c5afe5c5afafd3cae84e91. There's a small thread safety bug in the second commit (see below), but the previous concern I had about crashes isn't really an issue, and the current implementation of cancellation here seems pretty good.

w0xlt reviewed (commented):
reACK https://github.com/bitcoin/bitcoin/pull/34184/changes/1ae88d1688958fe567a087ec591015ea832aa8f0

Sjors commented:
@plebhash can you take this for a spin?

plebhash commented:
> @plebhash can you take this for a spin?

I checked out this branch, compiled it, launched it on `testnet4` and hooked SRI Pool via IPC.

my `testnet4` datadir had been off for a few days, so I left it catch up while observing Pool logs.

there was a flood of `NewTemplate` + `SetNewPrevHash` messages, just like the behavior reported on https://github.com/bitcoin/bitcoin/issues/33994

Sjors commented:
This is exceedingly hard to test / debug, due to a combination of four issues:

1. getting a node in a "almost done syncing" state is tedious (the trick it to `kill -9` the node when it reaches the tip, so it has to start over)
2. the node will crash if the client ungraciously exists
3. the sv2 client doesn't know about the interrupt method yet (since this PR introduces it)
4. `createNewBlock` doesn't have a `context` argument (known issue) so a second connection quietly fails to establish (in case of a test sscript that tries to connect a second time)

Once I clean up the code, I'll publish a gist with some patches to make testing easier without increasing the scope of this PR.

Sjors committed:
mining: add cooldown argument to createNewBlock()

At startup, if the needs to catch up, connected mining clients will
receive a flood of new templates as new blocks are connected.

Fix this by adding a cooldown argument to createNewBlock(). When set
to true, block template creation is briefly paused while the best
header chain is ahead of the tip.

This wait only happens when the best header extends the current tip,
to ignore competing branches.

Additionally, cooldown waits for isInitialBlockDownload() to latch to
false, which happens when there is less than a day of blocks left to sync.

When cooldown is false createNewBlock() returns immediately. The argument
is optional, because many tests are negatively impacted by this
mechanism, and single miner signets could end up stuck if no block
was mined for a day.

The getblocktemplate RPC also opts out, because it would add a delay
to each call.

Fixes #33994

Sjors commented:
I moved `run_deprecated_mining_test` down to avoid running into #34284.

ryanofsky commented on src/node/miner.cpp:
```diff
@@ -452,6 +452,38 @@ std::optional<BlockRef> GetTip(ChainstateManager& chainman)
+    uint256 last_tip_hash{last_tip.hash};
+
+    while (const std::optional<int> remaining = chainman.BlocksAheadOfTip()) {
+        const int cooldown_seconds = std::clamp(*remaining, 3, 20);
```
In commit "mining: add cooldown argument to createNewBlock()" (617eb13ec178154bdfd1a2f41496912d87bf17fc)

Like this, this is clever!

ryanofsky reviewed (approved):
Code review ACK 299f15677b3d32b878f79945128c977dc8eaedfd. Since last review, `createNewBlock` was fixed (by adding context parameter) to be actually interruptible, and cooldown was made optional and off by default, wait for `chainman().IsInitialBlockDownload()` was added, and `CooldownIfHeadersAhead` internal logic was improved to wait longer when there are more blocks waiting to be connected, instead of returning as soon as more than 1 second elapses between connecting blocks.

All of these changes seem very nice. I left suggestions for keeping backwards compatibility and fixing a missing lock with I think would be good to apply but probably not critical.

Sjors commented on src/interfaces/init.h:
```diff
@@ -39,6 +39,7 @@ class Init
     virtual Ipc* ipc() { return nullptr; }
     virtual bool canListenIpc() { return false; }
     virtual const char* exeName() { return nullptr; }
+    virtual void makeMiningV2() { throw std::runtime_error("Old mining interface (@2) not supported. Please update your client!"); }
```
Responded in the main thread.

Sjors committed:
mining: add interrupt()

Both waitTipChanged() and createNewBlock() can take a long time to
return. Add a way for clients to interrupt them.

The new m_interrupt_mining is safely accessed with a lock on
m_tip_block_mutex, but it has no guard annotation. A more thorough
solution is discussed here:
https://github.com/bitcoin/bitcoin/pull/34184#discussion_r2743566474

Sjors commented on src/node/interfaces.cpp:
```diff
@@ -982,18 +982,23 @@ class MinerImpl : public Mining
             while (chainman().IsInitialBlockDownload()) {
                 maybe_tip = waitTipChanged(maybe_tip->hash, MillisecondsDouble{1000});
-                if (!maybe_tip) return {};
+                if (!maybe_tip || m_interrupt) return {};
```
Good catch, seems I ignored my own source code comment.

I renamed it to `m_interrupt_mining` and also renamed the `CooldownIfHeadersAhead` argument, because it got confusing.

Sjors commented on src/ipc/capnp/init.capnp:
```diff
@@ -19,5 +19,8 @@ using Mining = import "mining.capnp";
+    makeMining @3 (context :Proxy.Context) -> (result :Mining.Mining);
+
+    # DEPRECATED: no longer supported; server returns an error.
+    makeMiningV2 @2 () -> ();
```
Changed to `makeMiningDeprecatedV2`

ryanofsky reviewed (approved):
Code review ACK e629a8262c083d4d210689961a4bbeb41851c0af. Changes since last review were rebasing to fix test conflicts and renaming the interrupt variable and adding a missing lock around one access.

I still feel like it would be better to replace the first commit breaking compatibility with the simpler change suggested in https://github.com/bitcoin/bitcoin/pull/34184#discussion_r2770232149 which keeps compatibility.

But I can understand being in a hurry to break compatibility when dealing with bugs in older clients & the pace of review here.

So assuming you do want to keep the commit which breaks compatibility (eaa1209ea4462d1221e049891fb588adb5cf60d3), I'd suggest two more changes from my notes about things to fix in the next version bump. One would be to credit @plebhash for first pointing out the missing Context arguments in his comment https://github.com/bitcoin/bitcoin/issues/33575#issuecomment-3383006863. I think he was the first person to notice that. The other would be to fix inconsistent defaults between C++ and Cap'n Proto interfaces, since changing default values also breaks backwards compatibility and the current missing defaults can lead to unnecessary problems and boilerplate in clients. Assigning consistent defaults should be possible with:

<details><summary>diff</summary>
<p>

```diff
--- a/src/ipc/capnp/mining.capnp
+++ b/src/ipc/capnp/mining.capnp
@@ -12,11 +12,15 @@ using Proxy = import "/mp/proxy.capnp";
 $Proxy.include("interfaces/mining.h");
 $Proxy.includeTypes("ipc/capnp/mining-types.h");
 
+const maxMoney :Int64 = 2100000000000000;
+const maxDouble :Float64 = 1.7976931348623157e308;
+const defaultBlockReservedWeight :UInt32 = 8000;
+
 interface Mining $Proxy.wrap("interfaces::Mining") {
     isTestChain @0 (context :Proxy.Context) -> (result: Bool);
     isInitialBlockDownload @1 (context :Proxy.Context) -> (result: Bool);
     getTip @2 (context :Proxy.Context) -> (result: Common.BlockRef, hasResult: Bool);
-    waitTipChanged @3 (context :Proxy.Context, currentTip: Data, timeout: Float64) -> (result: Common.BlockRef);
+    waitTipChanged @3 (context :Proxy.Context, currentTip: Data, timeout: Float64 = .maxDouble) -> (result: Common.BlockRef);
     createNewBlock @4 (context :Proxy.Context, options: BlockCreateOptions, cooldown: Bool) -> (result: BlockTemplate);
     checkBlock @5 (context :Proxy.Context, block: Data, options: BlockCheckOptions) -> (reason: Text, debug: Text, result: Bool);
     interrupt @6 () -> ();
@@ -39,19 +43,19 @@ interface BlockTemplate $Proxy.wrap("interfaces::BlockTemplate") {
 }
 
 struct BlockCreateOptions $Proxy.wrap("node::BlockCreateOptions") {
-    useMempool @0 :Bool $Proxy.name("use_mempool");
-    blockReservedWeight @1 :UInt64 $Proxy.name("block_reserved_weight");
-    coinbaseOutputMaxAdditionalSigops @2 :UInt64 $Proxy.name("coinbase_output_max_additional_sigops");
+    useMempool @0 :Bool = true $Proxy.name("use_mempool");
+    blockReservedWeight @1 :UInt64 = .defaultBlockReservedWeight $Proxy.name("block_reserved_weight");
+    coinbaseOutputMaxAdditionalSigops @2 :UInt64 = 400 $Proxy.name("coinbase_output_max_additional_sigops");
 }
 
 struct BlockWaitOptions $Proxy.wrap("node::BlockWaitOptions") {
-    timeout @0 : Float64 $Proxy.name("timeout");
-    feeThreshold @1 : Int64 $Proxy.name("fee_threshold");
+    timeout @0 : Float64 = .maxDouble $Proxy.name("timeout");
+    feeThreshold @1 : Int64 = .maxMoney $Proxy.name("fee_threshold");
 }
 
 struct BlockCheckOptions $Proxy.wrap("node::BlockCheckOptions") {
-    checkMerkleRoot @0 :Bool $Proxy.name("check_merkle_root");
-    checkPow @1 :Bool $Proxy.name("check_pow");
+    checkMerkleRoot @0 :Bool = true $Proxy.name("check_merkle_root");
+    checkPow @1 :Bool = true $Proxy.name("check_pow");
 }
 
 struct CoinbaseTx $Proxy.wrap("node::CoinbaseTx") {
```
</p>
</details>

PR also seems fine it its current form, though and these are just suggestions.

Sjors commented:
> I'd suggest two more changes 

I'll amend the commit to give credit if I need to touch it again.

The additional interface changes are in different methods, so I prefer to do them as a followup PR. I'll also drop the previously deprected methods there.

ismaelsadeeq commented on src/interfaces/mining.h:
```diff
@@ -145,13 +145,15 @@ class Mining
+     * @param[in] cooldown wait for tip to be connected and IBD to complete.
+     *                     If the best header is ahead of the tip, wait for the
+     *                     tip to catch up. Recommended, except for regtest and
+     *                     signets with only one miner.
```
In "mining: add cooldown argument to createNewBlock()" 0d4cf7041af41e6371473f7df93301d7100cea3e

nit: no need to mention all the custom setups that don't need cooldown; highlight that we return immediately.

Sjors replied:
Even though the default in our c++ code is `false` (see above), the recommendation for real clients is `true`, with these two exceptions.

ismaelsadeeq commented on src/node/miner.cpp:
```diff
@@ -452,6 +452,38 @@ std::optional<BlockRef> GetTip(ChainstateManager& chainman)
+    uint256 last_tip_hash{last_tip.hash};
+
+    while (const std::optional<int> remaining = chainman.BlocksAheadOfTip()) {
+        const int cooldown_seconds = std::clamp(*remaining, 3, 20);
```
In "mining: add cooldown argument to createNewBlock()" 0d4cf7041af41e6371473f7df93301d7100cea3e

Three seconds minimum seems a bit much, IMO, should just be 1?

ismaelsadeeq replied:
It's heuristics anyway, no strong opinion here.

ismaelsadeeq reviewed (commented):
Concept ACK

I briefly looked at this, you could still mine on a stale tip even with this when your node stalls for a bit for any reason.

But I get the point, mining on a stale, is far better than waiting forever.

I have some suggestions below.

Sjors commented on src/interfaces/mining.h:
```diff
@@ -145,13 +145,15 @@ class Mining
      * @retval std::nullptr if the node is shut down.
      */
-    virtual std::unique_ptr<BlockTemplate> createNewBlock(const node::BlockCreateOptions& options = {}) = 0;
+    virtual std::unique_ptr<BlockTemplate> createNewBlock(const node::BlockCreateOptions& options = {}, bool cooldown = false) = 0;
```
Ok, that's useful to know. I'll take a look at this PR. Will hold on to `false` until that lands though.

Sjors commented on src/node/miner.cpp:
```diff
@@ -452,6 +452,38 @@ std::optional<BlockRef> GetTip(ChainstateManager& chainman)
+
+    while (const std::optional<int> remaining = chainman.BlocksAheadOfTip()) {
+        const int cooldown_seconds = std::clamp(*remaining, 3, 20);
+        // Use a steady clock so cooldown is independent of mocktime.
```
Done!

ryanofsky reviewed (approved):
Code review ACK fe297f0976c1a2f8e2be3ad598bc287f71a7f8f9. Just rebased due to conflicts with https://github.com/bitcoin/bitcoin/pull/33965, added workaround to deal with new exception thrown by createNewBlock in that PR, added `cooldown=True` to python test, and switching from `SteadyClock` -> `MockableSteadyClock` since last review.

I'd like if #34568 could be merged before this PR so this PR could be cleaner and more narrowly focused on implementing the cooldown feature, and so the mining interface number wouldn't need to be bumped twice (or bumped once with an intermediate period where incompatible schemas could silently break clients).

I also think it would be good to add `cooldown: Bool = true` to the .capnp file if we think this is a better default behavior. It should be ok for capnp default value to be different from c++ if updating the c++ value would create too much churn in unit tests.

Just my thoughts though. PR is already in good shape as-is.

Sjors commented:
> I'd like if https://github.com/bitcoin/bitcoin/pull/34568 could be merged before this PR

That's fine by me.

> I also think it would be good to add `cooldown: Bool = true` to the `.capnp` file if we think this is a better default behavior.

Good idea. Since it's a new method, that's not a breaking change. It does make the diff bigger because I need to opt-out everywhere, but if we merge #34568 that reduces the PR diff.

Sjors committed:
doc: release note for IPC cooldown and interrupt

ryanofsky reviewed (approved):
Code review ACK 1e0288800e1f58a21462a07196c2a78de7b09518. Only changes since last review were switching cooldown default from false to true, rebasing, and adding release notes.

Might also be good to remove "Based on https://github.com/bitcoin/bitcoin/pull/34568" from the description to keep it simpler now the other PR is merged

ryanofsky reviewed (approved):
Code review ACK fcaec2544b32226fd5357a88506fe080058d25bc. Only changes since last review were removing two cooldown arguments from the mining IPC test to simplify it

enirox001 commented on src/node/interfaces.cpp:
```diff
@@ -978,18 +978,23 @@ class MinerImpl : public Mining
         return std::make_unique<BlockTemplateImpl>(assemble_options, BlockAssembler{chainman().ActiveChainstate(), context()->mempool.get(), assemble_options}.CreateNewBlock(), m_node);
     }
 
+    void interrupt() override
```
In commit "mining: add interrupt()" (1e82fa498c)

nano nit: The current naming is quite similar to `interruptWait()`. Explicitly documenting that this specifically targets the `createNewBlock` wait loop clarifies that these control two different internal primitives.

enirox001 reviewed (commented):
ACK fcaec2544b

Verified on Fedora 43 with a clean build. Functional tests pass including `interface_ipc_mining.py`

I have included a few suggestions and nits below

ismaelsadeeq reviewed (commented):
Post merge ACK fcaec2544b32226fd5357a88506fe080058d25bc
